fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/IndexingHeartbeatLowLevelTest.java

```diff
@@ -0,0 +1,473 @@
+        openMetaData(TestRecords1Proto.getDescriptor(), hook);
+    }
+
+    protected static FDBRecordStoreTestBase.RecordMetaDataHook allIndexesHook(List<Index> indexes) {
+        return metaDataBuilder -> {
+            for (Index index: indexes) {
+                metaDataBuilder.addIndex("MySimpleRecord", index);
+            }
+        } ;
+    }
+
+    @Test
+    void testHeartbeatQuery() {
+        List<Index> indexes = new ArrayList<>();
+        indexes.add(new Index("indexA", field("num_value_2"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS));
+        indexes.add(new Index("indexD", new GroupingKeyExpression(EmptyKeyExpression.EMPTY, 0), IndexTypes.COUNT));
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+
+        final int count = 23;
+        IndexingHeartbeat[] heartbeats = new IndexingHeartbeat[count];
+        for (int i = 0; i < count; i++) {
+            heartbeats[i] = new IndexingHeartbeat(UUID.randomUUID(), "heartbeat" + i, 100 + i, false);
+        }
+
+        // populate heartbeats
+        openSimpleMetaData(hook);
+        try (FDBRecordContext context = openContext()) {
+            for (var heartbeat : heartbeats) {
+                heartbeat.updateHeartbeat(recordStore, indexes.get(0));
+                heartbeat.updateHeartbeat(recordStore, indexes.get(1));
+            }
+            context.commit();
+        }
+
+        // Verify query operation
+        for (Index index: indexes) {
+            try (FDBRecordContext context = openContext()) {
+                // Query, unlimited
+                Map<UUID, IndexBuildProto.IndexBuildHeartbeat> queried =
+                        IndexingHeartbeat.getIndexingHeartbeats(recordStore, index, 0).join();
+                Assertions.assertThat(queried).hasSize(count);
+                Assertions.assertThat(queried.keySet())
+                        .containsExactlyInAnyOrderElementsOf(Arrays.stream(heartbeats).map(heartbeat -> heartbeat.indexerId).collect(Collectors.toList()));
+                Assertions.assertThat(queried.values().stream().map(IndexBuildProto.IndexBuildHeartbeat::getInfo))
+                        .containsExactlyInAnyOrderElementsOf(Arrays.stream(heartbeats).map(heartbeat -> heartbeat.info).collect(Collectors.toList()));
+                Assertions.assertThat(queried.values().stream().map(IndexBuildProto.IndexBuildHeartbeat::getGenesisTimeMilliseconds))
+                        .containsExactlyInAnyOrderElementsOf(Arrays.stream(heartbeats).map(heartbeat -> heartbeat.genesisTimeMilliseconds).collect(Collectors.toList()));
+
+                // Query, partial
+                queried =
+                        IndexingHeartbeat.getIndexingHeartbeats(recordStore, index, 5).join();
+                Assertions.assertThat(queried).hasSize(5);
+                context.commit();
+            }
+        }
+    }
+
+    @Test
+    void testHeartbeatLowLevelClearing() {
```

Comment:
Nit: This is no lower level than any of the other test in this class
```suggestion
    void testHeartbeatClearing() {
```

Reply:
√